### PR TITLE
UI keyboard and controller navigation

### PR DESCRIPTION
**Pull request by Jomosoto**
Allows navigating through the UI using the keyboard or a controller in the main menu, ingame and the editor pause menus:
- Grab UI focus
- Go back on "ui_cancel" input

Bugs to fix:
- [x] <del> Selecting a button in the ingame pause menu while a message is open closes the message </del>
- [x] <del> Controller navigation should be possible with the left stick, but `ItemList` doesn't support it</del> This is a bug in the Godot engine: https://github.com/godotengine/godot/issues/73241

### CODE REVIEW

File: src/Data/Scripts/StationJumper.gd [hunk 1]
```diff
@@ -7,6 +7,24 @@ var current_station_indices := [] # All indices of selectable stations
 signal station_index_selected(station_index)
 
 
+func _ready():
```
Comment (HaSa1002) [inline src/Data/Scripts/StationJumper.gd]: Please use typed GDScript. There's an editor setting for autocompletion to add types by default

```suggestion
func _ready() -> void:
```

File: src/Data/Scripts/StationJumper.gd [hunk 2]
```diff
@@ -7,6 +7,24 @@ var current_station_indices := [] # All indices of selectable stations
...
+	$StationJumper/ItemList.connect("item_activated", self, "_on_ItemList_item_activated")
+
+
+func show():
```
Comment (HaSa1002) [inline src/Data/Scripts/StationJumper.gd]:
```suggestion
func show() -> void:
```

File: src/Data/Scripts/StationJumper.gd [hunk 3]
```diff
@@ -7,6 +7,24 @@ var current_station_indices := [] # All indices of selectable stations
...
+	if $StationJumper/ItemList.get_item_count() > 0:
+		$StationJumper/ItemList.grab_focus()
+	else:
+		$StationJumper/HBoxContainer/Cancel.grab_focus()
+	.show()
+
+
+func _unhandled_input(event):
```
Comment (HaSa1002) [inline src/Data/Scripts/StationJumper.gd]:
```suggestion
func _unhandled_input(event: InputEvent) -> void:
```

File: src/Data/Scripts/StationJumper.gd [hunk 4]
```diff
@@ -7,6 +7,24 @@ var current_station_indices := [] # All indices of selectable stations
...
+	if event.is_action_pressed("ui_cancel") and visible:
```
Comment (HaSa1002) [inline src/Data/Scripts/StationJumper.gd]:
```suggestion
	if visible and event.is_action_pressed("ui_cancel"):
```

Please check properties first, especially if they're bools before calling methods

File: src/Data/Scripts/StationJumper.gd [hunk 5]
```diff
@@ -7,6 +7,24 @@ var current_station_indices := [] # All indices of selectable stations
 signal station_index_selected(station_index)
```
Comment (HaSa1002) [inline src/Data/Scripts/StationJumper.gd:10]: What about caching the item list and other UI controls?

File: src/Data/Scripts/StationJumper.gd [hunk 6]
```diff
@@ -7,6 +7,24 @@ var current_station_indices := [] # All indices of selectable stations
...
+		$StationJumper/ItemList.grab_focus()
+	else:
+		$StationJumper/HBoxContainer/Cancel.grab_focus()
+	.show()
+
+
+func _unhandled_input(event):
+	if event.is_action_pressed("ui_cancel") and visible:
+		_on_Cancel_pressed()
+		get_tree().set_input_as_handled()
```
Comment (HaSa1002) [inline src/Data/Scripts/StationJumper.gd]: Just FYI, you may get away with [`accept_event()`](https://docs.godotengine.org/en/stable/classes/class_control.html?highlight=Control#class-control-method-accept-event)

File: src/Data/Scripts/StationJumper.gd [hunk 7]
```diff
@@ -29,3 +48,7 @@ func _on_Jump_pressed() -> void:
 
 func _on_Cancel_pressed() -> void:
 	hide()
+
+
+func _on_ItemList_item_activated(index):
```
Comment (HaSa1002) [inline src/Data/Scripts/StationJumper.gd]:
```suggestion
func _on_ItemList_item_activated(index: int) -> void:
```

File: src/Data/UI/Content.gd [hunk 8]
```diff
@@ -10,9 +10,16 @@ func _ready() -> void:
 
 func show() -> void:
 	update_content_list()
+	$VBoxContainer/Buttons/Back.grab_focus()
 	.show()
 
 
+func _unhandled_input(event):
+	if event.is_action_pressed("ui_cancel") and visible:
```
Comment (HaSa1002) [inline src/Data/UI/Content.gd]:
```suggestion
func _unhandled_input(event: InputEvent) -> void:
	if visible and event.is_action_pressed("ui_cancel"):
```

File: src/Data/UI/main_menu.gd [hunk 9]
```diff
@@ -32,6 +35,44 @@ func _ready():
...
+	$Buttons/Content.connect("mouse_entered", self, "focus_button", ["Content"])
+	$Buttons/Create.connect("mouse_entered", self, "focus_button", ["Create"])
+	$Buttons/Settings.connect("mouse_entered", self, "focus_button", ["Settings"])
+	$Buttons/About.connect("mouse_entered", self, "focus_button", ["About"])
+	$Buttons/Feedback.connect("mouse_entered", self, "focus_button", ["Feedback"])
+	$Buttons/Quit.connect("mouse_entered", self, "focus_button", ["Quit"])
+	
+	$CreateMenu/Back.connect("mouse_entered", self, "createMenu_focus_button", ["Back"])
+	$CreateMenu/TrackEditor.connect("mouse_entered", self, "createMenu_focus_button", ["TrackEditor"])
+	$CreateMenu/ScenarioEditor.connect("mouse_entered", self, "createMenu_focus_button", ["ScenarioEditor"])
```
Comment (HaSa1002) [inline src/Data/UI/main_menu.gd]: Please don't. Focus should only set by mice when pressing the button. That's the default behaviour.
```suggestion
```

File: src/Data/UI/main_menu.gd [hunk 10]
```diff
@@ -32,6 +35,44 @@ func _ready():
...
+
+
+func focus_button(name):
+	$Buttons.get_node(name).grab_focus()
```
Comment (HaSa1002) [inline src/Data/UI/main_menu.gd]:
```suggestion
func focus_button(button: Button):
	button.grab_focus()
```

Use a node reference instead. It won't break the same way strings do.

File: src/Data/UI/main_menu.gd [hunk 11]
```diff
@@ -32,6 +35,44 @@ func _ready():
...
+
+
+func _unhandled_input(event):
+	if event.is_action_pressed("ui_cancel") and $CreateMenu.visible:
```
Comment (HaSa1002) [inline src/Data/UI/main_menu.gd]:
```suggestion
func _unhandled_input(event: InputEvent) -> void:
	if $CreateMenu.visible and event.is_action_pressed("ui_cancel"):
```

File: src/Data/UI/main_menu.gd [hunk 12]
```diff
@@ -85,5 +126,14 @@ func _on_CreateMenu_Back_pressed():
 	$Buttons.show()
 
 
+func _on_CreateMenu_visibility_changed():
+	if $CreateMenu.visible:
+		$CreateMenu/Back.grab_focus()
+
+
+func createMenu_focus_button(button_name):
+	$CreateMenu.get_node(button_name).grab_focus()
```
Comment (HaSa1002) [inline src/Data/UI/main_menu.gd]: note needed with change above.
```suggestion
```

File: src/Data/UI/main_menu.gd [hunk 13]
```diff
@@ -85,5 +126,14 @@ func _on_CreateMenu_Back_pressed():
 	$Buttons.show()
 
 
+func _on_CreateMenu_visibility_changed():
+	if $CreateMenu.visible:
```
Comment (HaSa1002) [inline src/Data/UI/main_menu.gd]:
```suggestion
func _on_CreateMenu_visibility_changed() -> void:
	if $CreateMenu.visible:
```
What about caching create menu?

File: src/Data/UI/menu_about.gd [hunk 14]
```diff
@@ -19,6 +19,17 @@ func _ready() -> void:
 	show_translators()
 
 
+func show():
```
Comment (HaSa1002) [inline src/Data/UI/menu_about.gd]:
```suggestion
func show() -> void:
```

File: src/Data/UI/menu_about.gd [hunk 15]
```diff
@@ -19,6 +19,17 @@ func _ready() -> void:
...
+	$Control/Back.grab_focus()
+	.show()
+
+
+func _unhandled_input(event):
+	if event.is_action_pressed("ui_cancel") and visible:
```
Comment (HaSa1002) [inline src/Data/UI/menu_about.gd]:
```suggestion
func _unhandled_input(event: InputEvent) -> void:
	if visible and event.is_action_pressed("ui_cancel"):
```

File: src/Data/UI/pause.gd [hunk 16]
```diff
@@ -10,8 +10,25 @@ var _saved_mouse_mode: int = 0
...
+func _ready():
+	$StationJumper.connect("hide", self, "focus_button", ["JumpToStation"])
+	jSettings.get_node("JSettings").connect("hide", self, "focus_button", ["Settings"])
+
+	$CenterContainer/HBox/Back.connect("mouse_entered", self, "focus_button", ["Back"])
+	$CenterContainer/HBox/RestartScenario.connect("mouse_entered", self, "focus_button", ["RestartScenario"])
+	$CenterContainer/HBox/JumpToStation.connect("mouse_entered", self, "focus_button", ["JumpToStation"])
+	$CenterContainer/HBox/Settings.connect("mouse_entered", self, "focus_button", ["Settings"])
+	$CenterContainer/HBox/QuitMenu.connect("mouse_entered", self, "focus_button", ["QuitMenu"])
+	$CenterContainer/HBox/QuitOS.connect("mouse_entered", self, "focus_button", ["QuitOS"])
```
Comment (HaSa1002) [inline src/Data/UI/pause.gd]:
```suggestion
```

File: src/Data/UI/pause.gd [hunk 17]
```diff
@@ -10,8 +10,25 @@ var _saved_mouse_mode: int = 0
...
+
+
+func show():
```
Comment (HaSa1002) [inline src/Data/UI/pause.gd]:
```suggestion
func show() -> void:
```

File: src/Data/UI/pause.gd [hunk 18]
```diff
@@ -10,8 +10,25 @@ var _saved_mouse_mode: int = 0
...
+	$CenterContainer/HBox/Back.grab_focus()
+	.show()
+
+
 func _unhandled_input(_event) -> void:
-	if Input.is_action_just_pressed("Escape"):
+	if Input.is_action_just_pressed("pause") and not jSettings.get_node("JSettings").visible:
```
Comment (HaSa1002) [inline src/Data/UI/pause.gd]: Can't you ask `jSettings` for its visible state?

File: src/Data/UI/pause.gd [hunk 19]
```diff
@@ -22,24 +39,29 @@ func _unhandled_input(_event) -> void:
 		elif not get_tree().paused:
 			Root.set_game_pause("ingame_pause", true)
 			jEssentials.show_message(tr("PAUSE_MODE_ENABLED"))
+	# Prevent "ui_accept" event from closing a potential message behind the menu
+	if Input.is_action_just_released("ui_accept") and visible:
```
Comment (HaSa1002) [inline src/Data/UI/pause.gd]:
```suggestion
	if visible and Input.is_action_just_released("ui_accept"):
		# Prevent "ui_accept" event from closing a potential message behind the menu
```
Personal preference: I'd move comments describing a block into the block.

File: src/Data/UI/pause.gd [hunk 20]
```diff
@@ -71,3 +93,7 @@ func _on_RestartScenario_pressed() -> void:
 
 func _on_Settings_pressed() -> void:
 	jSettings.popup()
+
+
+func focus_button(button_name):
+	$CenterContainer/HBox.get_node(button_name).grab_focus()
```
Comment (HaSa1002) [inline src/Data/UI/pause.gd]:
```suggestion
func focus_button(button: Button) -> void:
	button.grab_focus()
```

File: src/Data/UI/play_menu.gd [hunk 21]
```diff
@@ -10,11 +10,38 @@ var world_config: WorldConfig # config of selected world
...
+
+
 func show() -> void:
 	update_tracks()
+	$V/Tracks/H/ItemList.grab_focus()
 	.show()
 
 
+func _unhandled_input(event):
+	if event.is_action_pressed("ui_cancel") and visible:
```
Comment (HaSa1002) [inline src/Data/UI/play_menu.gd]:
```suggestion
func _unhandled_input(event: InputEvent) -> void:
	if visible and event.is_action_pressed("ui_cancel"):
```

File: src/Data/UI/play_menu.gd [hunk 22]
```diff
@@ -10,11 +10,38 @@ var world_config: WorldConfig # config of selected world
 var loaded_scenario: TrackScenario = null
 var loaded_route: ScenarioRoute = null
 
+
+func _ready():
```
Comment (HaSa1002) [inline src/Data/UI/play_menu.gd]:
```suggestion
func _ready() -> void:
```

File: src/Data/UI/play_menu.gd [hunk 23]
```diff
@@ -286,3 +317,16 @@ func _on_Trains_Play_pressed():
 
 func _on_Trains_item_activated(_index):
 	_on_Trains_Play_pressed()
+
+
+func _on_menu_visibility_changed():
```
Comment (HaSa1002) [inline src/Data/UI/play_menu.gd]:
```suggestion
func _on_menu_visibility_changed() -> void:
```

File: src/Data/UI/styles/button_both_focus.tres [hunk 24]
```diff
@@ -3,12 +3,11 @@
...
 content_margin_left = 38.0
 content_margin_right = 38.0
-bg_color = Color( 0.929412, 0.929412, 0.929412, 0.956863 )
-draw_center = false
+bg_color = Color( 0.764706, 0.890196, 0.921569, 0.956863 )
 border_width_left = 4
 border_width_top = 4
 border_width_right = 4
-border_color = Color( 0.2, 0.733333, 1, 0.486275 )
+border_color = Color( 0.164706, 0.6, 0.819608, 1 )
```
Comment (HaSa1002) [inline src/Data/UI/styles/button_both_focus.tres:10]: What's that? Wasn't the focus distinguishable enough?

File: src/Editor/Modules/scenario_editor_selection.tscn [hunk 25]
```diff
@@ -43,7 +43,6 @@ margin_top = 49.0
 margin_right = 1004.0
 margin_bottom = 527.0
 size_flags_vertical = 3
-select_mode = 1
```
Comment (HaSa1002) [inline src/Editor/Modules/scenario_editor_selection.tscn:46]: That's the track selection in the scenario editor menu, right?

File: src/Editor/Scripts/editor_configuration.gd [hunk 26]
```diff
@@ -10,6 +10,21 @@ func _ready() -> void:
...
+func show():
+	if tracks.empty():
+		$PanelContainer/VBoxContainer/TracksList/VBoxContainer/HBoxContainer/Back.grab_focus()
+	else:
+		$PanelContainer/VBoxContainer/TracksList/VBoxContainer/ItemList.grab_focus()
+	.show()
+
+
+func _unhandled_input(event):
+	if event.is_action_pressed("ui_cancel") and visible:
```
Comment (HaSa1002) [inline src/Editor/Scripts/editor_configuration.gd]:
```suggestion
func _unhandled_input(event: InputEvent) -> void:
	if visible and event.is_action_pressed("ui_cancel"):
```

File: src/Editor/Scripts/pause_menu.gd [hunk 27]
```diff
@@ -3,6 +3,17 @@ extends Control
 signal save_requested
 
 
+func _ready():
+	$VBoxContainer/Back.connect("mouse_entered", self, "_on_button_mouse_entered", ["Back"])
+	$VBoxContainer/ExitWithoutSave.connect("mouse_entered", self, "_on_button_mouse_entered", ["ExitWithoutSave"])
+	$VBoxContainer/SaveAndExit.connect("mouse_entered", self, "_on_button_mouse_entered", ["SaveAndExit"])
```
Comment (HaSa1002) [inline src/Editor/Scripts/pause_menu.gd]:
```suggestion
```

File: src/Editor/Scripts/pause_menu.gd [hunk 28]
```diff
@@ -28,3 +39,7 @@ func _on_hide() -> void:
 func _on_ExitWithoutSave_pressed() -> void:
 	get_tree().paused = false
 	LoadingScreen.load_main_menu()
+
+
+func _on_button_mouse_entered(button_name):
+	$VBoxContainer.get_node(button_name).grab_focus();
```
Comment (HaSa1002) [inline src/Editor/Scripts/pause_menu.gd]:
```suggestion
```

File: src/Editor/Scripts/scenario_editor.gd [hunk 29]
```diff
@@ -22,6 +22,15 @@ func _ready():
...
 
+	$CanvasLayer/Pause.connect("visibility_changed", self, "_on_Pause_visibility_changed")
+
+	$CanvasLayer/Pause/VBoxContainer/Back.connect("mouse_entered", self, "_on_Pause_button_mouse_entered", ["Back"])
+	$CanvasLayer/Pause/VBoxContainer/TestTrack.connect("mouse_entered", self, "_on_Pause_button_mouse_entered", ["TestTrack"])
+	$CanvasLayer/Pause/VBoxContainer/ExportTrack.connect("mouse_entered", self, "_on_Pause_button_mouse_entered", ["ExportTrack"])
+	$CanvasLayer/Pause/VBoxContainer/Save.connect("mouse_entered", self, "_on_Pause_button_mouse_entered", ["Save"])
+	$CanvasLayer/Pause/VBoxContainer/QuitWithoutSaving.connect("mouse_entered", self, "_on_Pause_button_mouse_entered", ["QuitWithoutSaving"])
+	$CanvasLayer/Pause/VBoxContainer/SaveAndQuit.connect("mouse_entered", self, "_on_Pause_button_mouse_entered", ["SaveAndQuit"])
+
```
Comment (HaSa1002) [inline src/Editor/Scripts/scenario_editor.gd]:
```suggestion
```

File: src/Editor/Scripts/scenario_editor.gd [hunk 30]
```diff
@@ -69,6 +78,15 @@ func _on_Pause_SaveAndQuit_pressed():
...
 
+func _on_Pause_visibility_changed():
+	if $CanvasLayer/Pause.visible:
+		$CanvasLayer/Pause/VBoxContainer/Back.grab_focus()
+
+
+func _on_Pause_button_mouse_entered(button_name):
+	$CanvasLayer/Pause/VBoxContainer.get_node(button_name).grab_focus()
+
+
```
Comment (HaSa1002) [inline src/Editor/Scripts/scenario_editor.gd]:
```suggestion
```

File: src/Editor/Scripts/scenario_editor_selection.gd [hunk 31]
```diff
@@ -2,6 +2,25 @@ extends Panel
...
+	$ScenarioList.connect("visibility_changed", self, "_on_ScenarioList_visibility_changed")
+
+
+func show():
+	$TrackList/ItemList.grab_focus()
+	.show()
+
+
+func _unhandled_input(event):
+	if event.is_action_pressed("ui_cancel") and visible:
```
Comment (HaSa1002) [inline src/Editor/Scripts/scenario_editor_selection.gd]:
```suggestion
func _unhandled_input(event: InputEvent) -> void:
	if visible and event.is_action_pressed("ui_cancel"):
```

File: src/Editor/Scripts/scenario_editor_selection.gd [hunk 32]
```diff
@@ -2,6 +2,25 @@ extends Panel
 
 var selected_track: String = ""
 
+
+func _ready():
+	$ScenarioList.connect("visibility_changed", self, "_on_ScenarioList_visibility_changed")
+
+
+func show():
```
Comment (HaSa1002) [inline src/Editor/Scripts/scenario_editor_selection.gd]:
```suggestion
func show() -> void:
```

File: src/Editor/Scripts/scenario_editor_selection.gd [hunk 33]
```diff
@@ -2,6 +2,25 @@ extends Panel
 
 var selected_track: String = ""
 
+
+func _ready():
```
Comment (HaSa1002) [inline src/Editor/Scripts/scenario_editor_selection.gd]:
```suggestion
func _ready() -> void:
```

File: src/addons/jean28518.jTools/jList/jList.gd [hunk 34]
```diff
@@ -109,6 +109,12 @@ func _unhandled_key_input(_event: InputEventKey) -> void:
 		_on_Add_pressed()
 
 
+func _unhandled_input(event):
+	if Input.is_action_just_pressed("ui_cancel") and $ConfirmationDialog.visible:
```
Comment (HaSa1002) [inline src/addons/jean28518.jTools/jList/jList.gd]:
```suggestion
func _unhandled_input(event: InputEvent) -> void:
	if $ConfirmationDialog.visible and Input.is_action_just_pressed("ui_cancel"):
```

**Review by HaSa1002 — CHANGES_REQUESTED** [review]
That's a pure code review. I haven't tested the PR yet. Please use typed GDScript whenever possible. There's also an editor setting to use typed GDScript in code completions.
The other issue I have is setting the focus on mouse enter. I understand why you did that, but I don't feel like that's common practice. Anything else is in the review comments. I'll give it another shot once it's finished. I'll do a functionality test then.

File: src/Editor/Modules/scenario_editor_selection.tscn (hunk 25 above)
Comment (Jomosoto) [reply src/Editor/Modules/scenario_editor_selection.tscn:46]: Yes, it was set to muti select so you can't move the selection by going up and down

File: src/Data/Scripts/StationJumper.gd (hunk 5 above)
Comment (Jomosoto) [reply src/Data/Scripts/StationJumper.gd:10]: What do you mean by caching?

**Comment by Jomosoto** [comment]
> The other issue I have is setting the focus on mouse enter. I understand why you did that, but I don't feel like that's common practice

I thought this is better for non-keyboard users, because otherwise there would be two highlighted buttons in every menu.

File: src/Data/Scripts/StationJumper.gd (hunk 5 above)
Comment (HaSa1002) [reply src/Data/Scripts/StationJumper.gd:10]: onready var node: ControlWhatEver = $What/Ever/Control

**Comment by HaSa1002** [comment]
Hmm. I see. I'd argue, we'd better hide the mouse and prevent hovered states that way?

**Comment by Jomosoto** [comment]
Sorry, I mean for those who use the mouse, there's always another item focused.

**Comment by HaSa1002** [comment]
Yeah but that's how every other UI works across all OS'

You could grab focus if none is focused and a keyboard/gamepad event is fired

File: src/Data/UI/main_menu.gd [hunk 35]
```diff
@@ -32,6 +35,44 @@ func _ready():
 	Logger.vlog("Main menu loaded")
 
 
+	# Signal connections for UI focus
+
+	$CreateMenu.connect("visibility_changed", self, "_on_CreateMenu_visibility_changed")
+
+	$Feedback.connect("popup_hide", self, "focus_button", ["Play"])
```
Comment (Jomosoto) [inline src/Data/UI/main_menu.gd]:
```suggestion
	$Feedback.connect("popup_hide", $Buttons/Play, "grab_focus")
```
Is this better?

Comment (HaSa1002) [reply src/Data/UI/main_menu.gd]: That's an option, too 😅

File: src/Data/UI/pause.gd (hunk 18 above)
Comment (Jomosoto) [reply src/Data/UI/pause.gd]: jSettings is always visible, only the visibility of jSettings/JSettings is changed.

File: src/Data/UI/styles/button_both_focus.tres (hunk 24 above)
Comment (Jomosoto) [reply src/Data/UI/styles/button_both_focus.tres:10]: No, there was just a small blue line around the button

**Comment by Jomosoto** [comment]
I'll fixup the commits called "Fixup" into the corresponding commits before this is merged. It's almost done, I just have to do the caching.

File: src/Data/Scripts/StationJumper.gd [hunk 36]
```diff
@@ -3,29 +3,54 @@ extends Control
 
 var current_station_indices := [] # All indices of selectable stations
 
+onready var itemList: ItemList = $StationJumper/ItemList
```
Comment (HaSa1002) [inline src/Data/Scripts/StationJumper.gd]: For the future, use snake_case for node references. [See the offical style guide.](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html#functions-and-variables) I am not going to make that a blocker though. So fix if you like.

File: src/Data/UI/pause.gd [hunk 37]
```diff
@@ -9,37 +9,53 @@ var _saved_ingame_pause: bool = false
...
+	.show()
+
 
 func _unhandled_input(_event) -> void:
-	if Input.is_action_just_pressed("Escape"):
-		if visible:
-			unpause()
-		else:
-			pause()
+	if visible and not settings.visible and (Input.is_action_just_pressed("pause") or Input.is_action_just_pressed("ui_cancel")):
```
Comment (HaSa1002) [inline src/Data/UI/pause.gd]:
```suggestion
	if visible and not settings.visible and \
			(Input.is_action_just_pressed("pause") \
			or Input.is_action_just_pressed("ui_cancel")):
```

FYI, you can add line breaks with the backslash to limit the line length.

**Review by HaSa1002 — COMMENTED** [review]
code review done. I'll continue with a test later this evening or tomorrow

**Review by HaSa1002 — CHANGES_REQUESTED** [review]
Please change the controller inputs to "All Devices". That's because the device handling is sadly a bit bad here when having multiple controller connected. I am going through alle the menus I can find now

**Review by HaSa1002 — CHANGES_REQUESTED** [review]
Here's a quick test review:
- [x] You can't move through the tabs in the settings
- [x] You can't scroll (may need an autoscroll)
- [x] Dunno if because of stick and list views but you there are two items focused in the play menu and the stick just ignores the list view
- [x] The start message button is not focused (A/Enter will work anyway because of the custom code there)
- [ ] Btw. we can't close doors on a controller 🤣  (not your bug)
- [x] Pressing the menu button sometimes closes the pause menu immediately again
- [x] Pressing Back (B) ingame while in some menu also closes the pause menu

**Comment by Jomosoto** [comment]
> You can't move through the tabs in the settings

Done, I just have to add controller icons. I took L1 and R1 to switch between tabs on controller and Page up and Page down on keyboard. (I wanted to take ctrl + left/right, but when focused the sliders steal the left/right input even if there are modifiers like ctrl. This is changed in Godot 4 if I remember correctly). Maybe I would only show the icons on controller, to not disturb mouse users(?). Is this fine in all?

> You can't scroll (may need an autoscroll)

> The start message button is not focused (A/Enter will work anyway because of the custom code there)

Done

> Dunno if because of stick and list views but you there are two items focused in the play menu and the stick just ignores the list view

I'm not sure what you mean. Because of the bug you can't change the selection in the ItemLists, but it's possible to focus the item list. I couldn't find where two items are focused; can you please send a screenshot?

> Pressing the menu button sometimes closes the pause menu immediately again

I couldn't reproduce. Can you please try in what conditions this happens or look why this happens?

> Pressing Back (B) ingame while in some menu also closes the pause menu

Hm, I can't reproduce. Do you mean pressing B while for example in the station jumper menu? For me this just closes the station jumper.

**Comment by HaSa1002** [comment]
> Maybe I would only show the icons on controller, to not disturb mouse users(?). Is this fine in all?

If you feel like adding Icons, go ahead but it's not gonna be an easy task. Could be done in a later PR.

> I'm not sure what you mean. Because of the bug you can't change the selection in the ItemLists, but it's possible to focus the item list. I couldn't find where two items are focused; can you please send a screenshot?

You can't move the focus with the left stick in the list. But not too much of an issue. Looks like a Godot issue to me.


> >Pressing the menu button sometimes closes the pause menu immediately again
>
> I couldn't reproduce. Can you please try in what conditions this happens or look why this happens?
> > Pressing Back (B) ingame while in some menu also closes the pause menu
>
> Hm, I can't reproduce. Do you mean pressing B while for example in the station jumper menu? For me this just closes the station jumper.

Yeah can't repro either. So seems to be not an issue anymore.

**Review by HaSa1002 — DISMISSED** [review]
lgtm

**Comment by HaSa1002** [comment]
I'll leave that unmerged for now. Tell me if you wanna continue here with the icons, but I'd merge without them. I think this PR is a great improvement and one of the last steps towards another beta.

**Comment by Jomosoto** [comment]
Thanks. I just have to fix the problem, that you can't see if a check_button is focused. I'll do the icons in a later PR.

**Review by HaSa1002 — APPROVED** [review]
lgtm

**Comment by HaSa1002** [comment]
Thank you very much ❤️

**Pull request merged**